# Data Cleaning with Pandas

In [46]:
import pandas as pd
import numpy as np

## Scenario

As data scientists, we want to build a model to predict the sale price of a house in Seattle in 2019, based on its square footage. We know that the King County Department of Assessments has comprehensive data available on real property sales in the Seattle area. We need to prepare the data.

### First, get the data!

When working on a project involving data that can fit on our computer, we store it in a `data` directory.

```bash
cd <project_directory>  # example: cd ~/flatiron_ds/pandas-3
mkdir data
cd data
```

Note that `<project_directory>` in angle brackets is a _placeholder_. You should type the path to the actual location on your computer where you're working on this project. Do not literally type `<project_directory>` and _do not type the angle brackets_. You can see an example in the _comment_ to the right of the command above.

Now, we'll need to download the two data files that we need. We can do this at the command line:

```bash
wget https://aqua.kingcounty.gov/extranet/assessor/Real%20Property%20Sales.zip
wget https://aqua.kingcounty.gov/extranet/assessor/Residential%20Building.zip
```

*Note:* If you do not have the `wget` command yet, you can install it: `brew install wget`.

Note that `%20` in a URL translates into a space. Even though you will *never put spaces in filenames*, you may need to deal with spaces that _other_ people have used in filenames.

There are two ways to handle the spaces in these filenames when referencing them at the command line.

#### 1. You can _escape_ the spaces by putting a backslash (`\`, remember _backslash is next to backspace_) before each one:

`unzip Real\ Property\ Sales.zip`

This is what happens if you tab-complete the filename in the terminal. Tab completion is your friend!

#### 2. You can put the entire filename in quotes:

`unzip "Real Property Sales.zip"`

Try unzipping these files with the `unzip` command. The `unzip` command takes one argument, the name of the tile that you want to unzip.


You can use tab completion (press the `tab` key after the first three letters) to fill in the names, including spaces. This will 

In [47]:
sales_df = pd.read_csv('data/EXTR_RPSale.csv')

/anaconda3/envs/learn-env/lib/python3.6/site-packages/IPython/core/interactiveshell.py:2785: DtypeWarning: Columns (1,2) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


### Seeing pink? Warnings are useful!

Note the warning above: `DtypeWarning: Columns (1, 2) have mixed types.` Because we start with an index of zero, the columns that we're being warned about are actually the _second_ and _third_ columns, `sales_df['Major']` and `sales_df['Minor']`.

In [48]:
sales_df.columns

Index(['ExciseTaxNbr', 'Major', 'Minor', 'DocumentDate', 'SalePrice',
       'RecordingNbr', 'Volume', 'Page', 'PlatNbr', 'PlatType', 'PlatLot',
       'PlatBlock', 'SellerName', 'BuyerName', 'PropertyType', 'PrincipalUse',
       'SaleInstrument', 'AFForestLand', 'AFCurrentUseLand', 'AFNonProfitUse',
       'AFHistoricProperty', 'SaleReason', 'PropertyClass', 'SaleWarning'],
      dtype='object')

what information do I want to use? what processes generated the data you're looking to use, what data was made using different processes?

In [49]:
sales_df.head().T

0  \
ExciseTaxNbr                                                  2687551   
Major                                                          138860   
Minor                                                             110   
DocumentDate                                               08/21/2014   
SalePrice                                                      245000   
RecordingNbr                                           20140828001436   
Volume                                                                  
Page                                                                    
PlatNbr                                                                 
PlatType                                                                
PlatLot                                                                 
PlatBlock                                                               
SellerName          WENKLE NOEL SMITH -TRUSTEE                    ...   
BuyerName           ALEXANDER APRIL                               ...   
PropertyType                                                        3   
PrincipalUse                                                        6   
SaleInstrument                                                      3   
AFForestLand                                                        N   
AFCurrentUseLand                                                    N   
AFNonProfitUse                                                      N   
AFHistoricProperty                                                  N   
SaleReason                                                          1   
PropertyClass                                                       8   
SaleWarning                                                             

                                                                    1  \
ExciseTaxNbr                                                  1235111   
Major                                                          664885   
Minor                                                              40   
DocumentDate                                               07/09/1991   
SalePrice                                                           0   
RecordingNbr                                           199203161090     
Volume                                                            071   
Page                                                              001   
PlatNbr                                                        664885   
PlatType                                                            C   
PlatLot                                                B102             
PlatBlock                                                               
SellerName          HERRMAN JEFFREY A+PAULINE A                   ...   
BuyerName           HERRMAN PAULINE A TTEE                        ...   
PropertyType                                                        3   
PrincipalUse                                                        0   
SaleInstrument                                                     26   
AFForestLand                                                        N   
AFCurrentUseLand                                                    N   
AFNonProfitUse                                                      N   
AFHistoricProperty                                                  N   
SaleReason                                                         18   
PropertyClass                                                       3   
SaleWarning                                                        11   

                                                                    2  \
ExciseTaxNbr                                                  2704079   
Major                                                          423943   
Minor                                                              50   
DocumentDate                                               10/11/2014   
SalePrice                                          

### Data overload?

That's a lot of columns. We're only interested in identifying the date, sale price, and square footage of each specific property. What can we do?

In [50]:
sales_df = sales_df[['Major', 'Minor', 'DocumentDate', 'SalePrice']]

In [51]:
sales_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2014336 entries, 0 to 2014335
Data columns (total 4 columns):
Major           object
Minor           object
DocumentDate    object
SalePrice       int64
dtypes: int64(1), object(3)
memory usage: 61.5+ MB


In [52]:
bldg_df = pd.read_csv('data/Residential Building.zip')

/anaconda3/envs/learn-env/lib/python3.6/site-packages/IPython/core/interactiveshell.py:2785: DtypeWarning: Columns (11) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


### Another warning! Which column has index 11?

In [53]:
bldg_df.columns[11]

'ZipCode'

`ZipCode` seems like a potentially useful column. We'll need it to determine which house sales took place in Seattle.

### So many features!

As data scientists, we should be _very_ cautious about discarding potentially useful data. But, today, we're interested in _only_ the total square footage of each property. What can we do?


In [54]:
bldg_df = bldg_df[['Major', 'Minor', 'SqFtTotLiving', 'ZipCode']]

In [55]:
bldg_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 511412 entries, 0 to 511411
Data columns (total 4 columns):
Major            511412 non-null int64
Minor            511412 non-null int64
SqFtTotLiving    511412 non-null int64
ZipCode          468372 non-null object
dtypes: int64(3), object(1)
memory usage: 15.6+ MB


In [56]:
#sales_data = pd.merge(sales_df, bldg_df, on=['Major', 'Minor'])

### Error!

Why are we seeing an error when we try to join the dataframes?

<table>
    <tr>
        <td style="text-align:left"><pre>
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2013160 entries, 0 to 2013159
Data columns (total 4 columns):
Major           object
Minor           object
DocumentDate    object
SalePrice       int64
dtypes: int64(1), object(3)
memory usage: 61.4+ MB</pre></td>
        <td style="text-align:left"><pre>
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 511359 entries, 0 to 511358
Data columns (total 4 columns):
Major            511359 non-null int64
Minor            511359 non-null int64
SqFtTotLiving    511359 non-null int64
ZipCode          468345 non-null object
dtypes: int64(3), object(1)
memory usage: 15.6+ MB
</pre></td>
    </tr>
</table>

Review the error message in light of the above:

* `ValueError: You are trying to merge on object and int64 columns.`

In [57]:
sales_df['Major'].head()

0    138860
1    664885
2    423943
3    403700
4    951120
Name: Major, dtype: object

### Error!

Note the useful error message above:

`ValueError: Unable to parse string "      " at position 936643`

In this case, we want to treat non-numeric values as missing values. Let's see if there's a way to change how the `pd.to_numeric` function handles errors.

In [58]:
# The single question mark means "show me the docstring"
#pd.to_numeric?

Here's the part that we're looking for:
```
errors : {'ignore', 'raise', 'coerce'}, default 'raise'
    - If 'raise', then invalid parsing will raise an exception
    - If 'coerce', then invalid parsing will be set as NaN
    - If 'ignore', then invalid parsing will return the input
```

Let's try setting the `errors` parameter to `'coerce'`.

In [59]:
sales_df['Major'] = pd.to_numeric(sales_df['Major'], errors='coerce')

Did it work?

In [60]:
sales_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2014336 entries, 0 to 2014335
Data columns (total 4 columns):
Major           float64
Minor           object
DocumentDate    object
SalePrice       int64
dtypes: float64(1), int64(1), object(2)
memory usage: 61.5+ MB


It worked! Let's do the same thing with the `Minor` parcel number.

In [61]:
sales_df['Minor'] = pd.to_numeric(sales_df['Minor'], errors='coerce')

In [62]:
sales_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2014336 entries, 0 to 2014335
Data columns (total 4 columns):
Major           float64
Minor           float64
DocumentDate    object
SalePrice       int64
dtypes: float64(2), int64(1), object(1)
memory usage: 61.5+ MB


Now, let's try our join again.

In [63]:
sales_data = pd.merge(sales_df, bldg_df, on=['Major', 'Minor'])

In [64]:
sales_data.head()

Major  Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode
0  138860.0  110.0   08/21/2014     245000           1490   98002
1  138860.0  110.0   06/12/1989     109300           1490   98002
2  138860.0  110.0   01/16/2005      14684           1490   98002
3  138860.0  110.0   06/08/2005          0           1490   98002
4  423943.0   50.0   10/11/2014          0            960   98092

In [65]:
sales_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1437522 entries, 0 to 1437521
Data columns (total 6 columns):
Major            1437522 non-null float64
Minor            1437522 non-null float64
DocumentDate     1437522 non-null object
SalePrice        1437522 non-null int64
SqFtTotLiving    1437522 non-null int64
ZipCode          1322148 non-null object
dtypes: float64(2), int64(2), object(2)
memory usage: 76.8+ MB


We can see right away that we're missing zip codes for many of the sales transactions. (1321536 non-null entries for ZipCode is fewer than the 1436772 entries in the dataframe.) 

In [66]:
sales_data.loc[sales_data['ZipCode'].isna()].head()

Major  Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode
99   858140.0  376.0   05/22/2012          0            900     NaN
100  858140.0  376.0   11/28/2017          0            900     NaN
152  720319.0  520.0   11/20/2013     699950           2840     NaN
153  720319.0  520.0   09/16/2013          0           2840     NaN
163  894677.0  240.0   12/21/2016     818161           2450     NaN

Because we are interested in finding houses in Seattle zip codes, we will need to drop the rows with missing zip codes.

In [67]:
sales_data = sales_data.loc[~sales_data['ZipCode'].isna(), :]

sales_data.head()

Major  Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode
0  138860.0  110.0   08/21/2014     245000           1490   98002
1  138860.0  110.0   06/12/1989     109300           1490   98002
2  138860.0  110.0   01/16/2005      14684           1490   98002
3  138860.0  110.0   06/08/2005          0           1490   98002
4  423943.0   50.0   10/11/2014          0            960   98092

# Your turn: Data Cleaning with Pandas

### 1. Investigate and drop rows with invalid values in the SalePrice and SqFtTotLiving columns.

Use multiple notebook cells to accomplish this! Press `[esc]` then `B` to create a new cell below the current cell. Press `[return]` to start typing in the new cell.

In [68]:
#find the names of the columns
sales_data.columns

Index(['Major', 'Minor', 'DocumentDate', 'SalePrice', 'SqFtTotLiving',
       'ZipCode'],
      dtype='object')

In [69]:
#see how many of the values in 'SalePrice' are not numbers
sales_data.loc[sales_data['SalePrice'].isna()].head()

Empty DataFrame
Columns: [Major, Minor, DocumentDate, SalePrice, SqFtTotLiving, ZipCode]
Index: []

In [70]:
#show me only the rows where the 'SalePrice' value is 0, then bring me the list of their indices
not_normal_home_sale = sales_data[sales_data['SalePrice']==0].index

# Delete these row indexes from dataFrame
sales_data.drop(not_normal_home_sale, inplace=True, axis=0)
sales_data.head()

Major  Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode
0  138860.0  110.0   08/21/2014     245000           1490   98002
1  138860.0  110.0   06/12/1989     109300           1490   98002
2  138860.0  110.0   01/16/2005      14684           1490   98002
6  423943.0   50.0   07/15/1999      96000            960   98092
7  423943.0   50.0   01/08/2001     127500            960   98092

### 2. Investigate and handle non-numeric ZipCode values

Can you find a way to shorten ZIP+4 codes to the first five digits?

What's the right thing to do with missing values?

In [71]:
sales_data.ZipCode.describe()

count     867416
unique       271
top        98042
freq       25815
Name: ZipCode, dtype: object

In [72]:
#confirm the right length of zip code we want
sales_data.ZipCode[0][0:5]

'98002'

In [73]:
#convert the series to strings
sales_data.ZipCode = sales_data.ZipCode.apply(np.str)

In [74]:
#strip away any empty space
sales_data.ZipCode.str.strip()

0            98002
1            98002
2            98002
6            98092
7            98092
11           98008
12           98008
13           98008
14           98008
16           98058
17           98058
19           98038
20           98038
21           98038
22           98038
23           98038
25           98031
26           98031
27           98031
29           98058
31           98058
32           98058
33           98058
34           98188
36           98188
38           98051
39           98051
40           98001
42           98001
44           98001
            ...   
1437478      98116
1437479      98146
1437481      98042
1437484      98125
1437485      98077
1437487      98031
1437488      98072
1437489      98053
1437490      98105
1437491      98102
1437492      98042
1437493      98042
1437498      98105
1437499      98005
1437500      98005
1437502    98115.0
1437504      98077
1437505      98001
1437508      98109
1437509      98058
1437510      98028
1437511     

In [75]:
#get the index numbers of zip codes that are 5 digits long
too_long_zips = sales_data[sales_data.ZipCode.str.len() > 5].index
#sales_data[sales_data['SalePrice']==0].index


In [76]:
#drop rows that don't have that criteria
sales_data.drop(too_long_zips, inplace=True, axis=0)
#not_normal_home_sale, inplace=True, axis=0

In [77]:
#confirm we've dealth with the too long zip codes
sales_data[sales_data.ZipCode.str.len() > 5]

Empty DataFrame
Columns: [Major, Minor, DocumentDate, SalePrice, SqFtTotLiving, ZipCode]
Index: []

In [78]:
#do the same with the too short zip codes
too_short_zips = sales_data[sales_data.ZipCode.str.len() < 5].index
sales_data.drop(too_short_zips, inplace=True, axis=0)

In [79]:
#confirm the too short zips are gone
sales_data[sales_data.ZipCode.str.len() < 5]

Empty DataFrame
Columns: [Major, Minor, DocumentDate, SalePrice, SqFtTotLiving, ZipCode]
Index: []

In [80]:
#convert the zip codes to ints
sales_data.ZipCode = sales_data.ZipCode.apply(np.int64)

In [81]:
#confirm all the zipcodes are the right length
sales_data.ZipCode.unique()

array([98002, 98092, 98008, 98058, 98038, 98031, 98188, 98051, 98001,
       98108, 98198, 98115, 98118, 98072, 98117, 98039, 98155, 98075,
       98003, 98103, 98022, 98042, 98040, 98133, 98105, 98056, 98102,
       98053, 98168, 98027, 98011, 98074, 98146, 98024, 98029, 98006,
       98005, 98028, 98034, 98144, 98030, 98177, 98166, 98065, 98112,
       98116, 98010, 98199, 98032, 98106, 98059, 98070, 98045, 98136,
       98125, 98023, 98033, 98077, 98109, 98055, 98178, 98052, 98122,
       98014, 98004, 98119, 98107, 98126, 98007, 98019, 98224, 98148,
       98047, 98288, 98050, 98354, 98068, 98057, 98302, 98083, 98121,
       98043, 98113, 98134, 98026, 89118, 98104, 98000, 98204, 98035,
       98422, 98132, 95059, 98097, 28028, 98189, 98017, 98025, 98044,
       98111, 90108, 98079, 98021, 95055, 98009, 98063, 98176, 98101,
       99019, 90855, 98054])

### 3. Add a column for PricePerSqFt



In [82]:
#add the price per sq foot column
sales_data['PricePerSqFt'] = sales_data.SalePrice / sales_data.SqFtTotLiving

### 4. Subset the data to 2019 sales only.

We can assume that the DocumentDate is approximately the sale date.

In [83]:
#only show the rows where the date has 2019 in it
sales_data[sales_data.DocumentDate.str.contains('2019')].head()

Major  Minor DocumentDate  SalePrice  SqFtTotLiving  ZipCode  \
605   214090.0  110.0   03/06/2019     390000           1780    98042   
651   367890.0   30.0   03/13/2019     923000           2550    98144   
868   350160.0  125.0   02/06/2019     935000           2460    98117   
1063  882790.0  870.0   02/15/2019     890000           2380    98105   
1341   30200.0  645.0   03/04/2019     380000           1750    98001   

      PricePerSqFt  
605     219.101124  
651     361.960784  
868     380.081301  
1063    373.949580  
1341    217.142857

### 5. Subset the data to zip codes within the City of Seattle.

You'll need to find a list of Seattle zip codes!

In [84]:
#make a list of seattle zip codes
seattle_city_zips = [98101, 98102, 98103, 98104, 98105, 98106, 98107, 98108, 98109, 98110, 98111, 98112, 98113, 98114, 98115, 
                     98116, 98117, 98118, 98119, 98121, 98122, 98124, 98125, 98126, 98127, 98129, 98131, 98133, 98134, 
                     98136, 98138, 98139, 98141, 98144, 98145, 98146, 98148, 98154, 98155, 98158, 98160, 98161, 98164, 
                     98165, 98166, 98168, 98170, 98174, 98175, 98177, 98178, 98181, 98185, 98188, 98190, 98191, 98194, 
                     98195, 98199]

In [85]:
#set this new dataframe to a variable
seattle_sales = sales_data.loc[sales_data.ZipCode.isin(seattle_city_zips)]


### 6. What is the mean price per square foot for a house sold in Seattle in 2019?

Don't just type the answer. Type code that generates the answer as output!

In [86]:
#exclude the impossible large numbers and the rediculously small numbers
seattle_sales = seattle_sales.loc[seattle_sales.PricePerSqFt <10000]
seattle_sales = seattle_sales.loc[seattle_sales.PricePerSqFt >0]

In [89]:
#lookup rows that have a document date with 2019 in it
#then lookup their price per square foot and find the mean
seattle_sales.loc[seattle_sales['DocumentDate'].str.contains('2019')].PricePerSqFt.mean()

480.0617565732615

In [88]:
#I'm done now!